File: rotate.java

// Time Complexity :O(n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no
class Solution {
    public void rotate(int[] nums, int k) {
        if(nums == null || nums.length == 0) return;
        int n = nums.length;
        if(k >= n){
            k = k%n;
        }
        reverse(nums, 0, n-1);
        reverse(nums, 0, k-1);
        reverse(nums, k,n-1);
    }
    private void reverse(int [] nums, int l, int r){
        while(l < r){
            swap(nums, l, r);
            l++; r--;
        }
    }
    private void swap(int [] arr, int i, int j){
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}

File: rotate.cpp
//TC: O(N)
//SC: O(1)

class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        
        
        if(nums.size() <= 1)
            return;
        
        
        int n = k%nums.size();
        
        reverse(nums.begin(), nums.end());
        reverse(nums.begin(), nums.begin()+n);
        reverse(nums.begin()+n, nums.end());
        
    }
};